docker/Makefile

```diff
@@ -0,0 +1,213 @@
+DOCKER_ARGS = $(DOCKER_BUILDARGS)
+
+IMAGES=contracts
+
+# for the most part this is just used to force rebuild when the
+# PDO contracts has changed
+TIMESTAMP := $(shell /bin/date "+%Y%m%d%H%M%S")
+
+all : $(addprefix build_,$(IMAGES))
+
+rebuild_% : contracts
+	@ docker build $(DOCKER_ARGS) \
+		--build-arg REBUILD=$(TIMESTAMP) \
+		--tag pdo_$*:$(CONTRACTS_VERSION) \
+		--file $(DOCKER_DIR)/pdo_$*.dockerfile .
+
+build_% : contracts
+	@ docker build $(DOCKER_ARGS) \
+		--tag pdo_$*:$(CONTRACTS_VERSION) \
+		--file $(DOCKER_DIR)/pdo_$*.dockerfile .
+
+# -----------------------------------------------------------------
+# -----------------------------------------------------------------
+# For the purposes of testing, we just use the resnet model for
+# classifying images. To use a different model, you should be able
+# to override these variables in mkae.loc
+OPENVINO_MODEL_BASE_URL ?= https://storage.openvinotoolkit.org/repositories/open_model_zoo/2022.1/models_bin/2
+OPENVINO_MODEL ?= resnet50-binary-0001
+
+models :
+	mkdir -p $(DOCKER_DIR)/models/1
+	wget -P $(DOCKER_DIR)/models/1 ${OPENVINO_MODEL_BASE_URL}/${OPENVINO_MODEL}/FP32-INT1/${OPENVINO_MODEL}.bin
+	wget -P $(DOCKER_DIR)/models/1 ${OPENVINO_MODEL_BASE_URL}/${OPENVINO_MODEL}/FP32-INT1/${OPENVINO_MODEL}.xml
+
+clean_models :
+	rm -rf $(DOCKER_DIR)/models
+
+OPENVINO_CONTAINER=openvino/model_server:latest
+OPENVINO_RUN_ARGS += -u $(CONTRACTS_USER_ID):$(CONTRACTS_GROUP_ID)
+OPENVINO_RUN_ARGS += -v $(DOCKER_DIR)/models/:/models/
+OPENVINO_RUN_ARGS += -p 127.0.0.1:9000:9000
+OPENVINO_RUN_COMMAND ?= --model_path /models/ --model_name resnet --port 9000 --shape auto
+
+run_openvino : models
+	docker run $(OPENVINO_RUN_ARGS) -d --name openvino_container $(OPENVINO_CONTAINER) \
+		$(OPENVINO_RUN_COMMAND)
+
+stop_openvino :
+	- docker rm -f openvino_container
+
+# -----------------------------------------------------------------
+# -----------------------------------------------------------------
+clean_% :
+	@ docker rmi -f pdo_$*
+
+# -----------------------------------------------------------------
+# -----------------------------------------------------------------
+DOCKER_RUN_ARGS += -v $(DOCKER_DIR)/xfer/:/project/pdo/xfer
+DOCKER_RUN_ARGS += --network host
+DOCKER_RUN_COMMAND ?= "stty cols $$(tput cols) rows $$(tput lines) && bash"
+
+stop_contracts :
+	- docker rm -f contracts_container
+
+run_contracts : build_contracts stop_contracts
+	docker run $(DOCKER_RUN_ARGS) -it --name contracts_container pdo_contracts:${CONTRACTS_VERSION} \
+		-c $(DOCKER_RUN_COMMAND)
+
+# -----------------------------------------------------------------
+# -----------------------------------------------------------------
+JUPYTER_SERVICE_HOST ?= $(PDO_HOSTNAME)
+JUPYTER_LEDGER ?= $(PDO_LEDGER_URL)
+JUPYTER_MODE ?= build
+
+DOCKER_JUPYTER_COMMAND = "/project/pdo/tools/start_jupyter.sh"
+DOCKER_JUPYTER_COMMAND += -l $(JUPYTER_LEDGER)
+DOCKER_JUPYTER_COMMAND += --service-host $(JUPYTER_SERVICE_HOST)
```

Comment:
I thought i could configure here the interface on which jupyter listens to (so i could access  headless remote test machine without having to do ssh port tunneling) but this seems actually to map to PDO_HOSTNAME and the infrastructure servers?)  Put differently, shouldn't there be (at least) two parameters, one for the interface jupyter is listening to and another one where the infrastructure is?

Reply:
you can. set JUPYTER_SERVICE_HOST in the make.loc. you cannot change the ports. we don't support that for PDO or for contracts.

that being said... you actually can override the ports (for services) if you really want to. the "--mode copy" switch for start_client.sh and start_jupyter.sh allows you to copy configuration files from xfer/client/etc. its pretty easy to create a pcontract.toml (or whatever you want for services). 

for changing the jupyter ports, no, you cannot currently do this from the command line (there is no switch for start_jupyter to set a port). this can be added though i think the value is pretty low.

Reply:
> you can. set JUPYTER_SERVICE_HOST in the make.loc. you cannot change the ports. we don't support that for PDO or for contracts.

I know that the makefile allows that and i also tried.  But it still seems to have jupyter listen on localhost/127.0.0.1. also from start_jupyter.sh i wouldn't see how that would influence to where jupyter listens too?  JUPYTER_SERVICE_HOST  translates to F_SERVICE_HOST which then is used only to define PDO_HOSTNAME and that doesn't seem to be used anywhere related to jupyter?

Reply:
oh... service host is where the eservices are running by default. sorry misunderstood. any reason why ssh tunneling isn't sufficient for your requirements?

Reply:
Well, ssh tunneling does work, just thought that variable would provide an easier path.  I guess just something to clarify in a README as discussed elsewhere?